neo/SmartContract/NativeContracts/GasToken.cs

```diff
@@ -0,0 +1,150 @@
+                Key = account
+            });
+            if (storage is null) return BigInteger.Zero;
+            Struct state = (Struct)storage.Value.DeserializeStackItem(engine.MaxArraySize);
+            return state[0].GetBigInteger();
+        }
+
+        private bool GasToken_Transfer(ExecutionEngine engine, byte[] from, byte[] to, BigInteger amount)
+        {
+            UInt160 hash_from = new UInt160(from);
+            UInt160 hash_to = new UInt160(to);
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (!hash_from.Equals(new UInt160(engine.CurrentContext.CallingScriptHash)) && !CheckWitness(engine, new UInt160(from)))
+                return false;
+            ContractState contract_to = Snapshot.Contracts.TryGet(hash_to);
+            if (contract_to?.Payable != true) return false;
+            if (amount.Sign > 0)
+            {
+                StorageKey key_from = new StorageKey
+                {
+                    ScriptHash = Blockchain.GasToken.ScriptHash,
+                    Key = from
+                };
+                StorageItem storage_from = Snapshot.Storages.TryGet(key_from);
+                if (storage_from is null) return false;
+                GasToken_AccountState state_from = GasToken_AccountState.FromStruct((Struct)storage_from.Value.DeserializeStackItem(engine.MaxArraySize));
+                if (state_from.Balance < amount) return false;
+                if (!hash_from.Equals(hash_to))
+                {
+                    if (state_from.Balance == amount)
+                    {
+                        Snapshot.Storages.Delete(key_from);
+                    }
+                    else
+                    {
+                        state_from.Balance -= amount;
+                        storage_from = Snapshot.Storages.GetAndChange(key_from);
+                        storage_from.Value = state_from.ToStruct().Serialize();
+                    }
+                    StorageKey key_to = new StorageKey
+                    {
+                        ScriptHash = Blockchain.GasToken.ScriptHash,
+                        Key = to
+                    };
+                    StorageItem storage_to = Snapshot.Storages.GetAndChange(key_to, () => new StorageItem
+                    {
+                        Value = new GasToken_AccountState().ToStruct().Serialize()
+                    });
+                    GasToken_AccountState state_to = GasToken_AccountState.FromStruct((Struct)storage_to.Value.DeserializeStackItem(engine.MaxArraySize));
+                    state_to.Balance += amount;
+                    storage_to.Value = state_to.ToStruct().Serialize();
+                }
+            }
+            SendNotification(engine, new StackItem[] { "Transfer", from, to, amount });
+            return true;
+        }
+
+        private class GasToken_AccountState
+        {
+            public BigInteger Balance;
```

Comment:
If we will store just the balance, maybe we should save the BigInteger without structs

Reply:
In the future, a standard for storing balances may be developed so that wallets/clients can read account balances directly from storage without calling the NEP-5 contracts. This standard, I am currently thinking about saving the balance in the first field of the struct, and then storing the serialized struct into the storage.